### ISSUE TITLE
HTTPS Everywhere licensing and manually editing Web pages to use HTTPS for links

### ISSUE THREAD

**Issue by galenide** [EFForg/https-everywhere#9571]
Something that I have done at times is manually editing Web pages (including, for example, pages on Wikipedia) to increase the extent to which the page's links use HTTPS. The process for editing a page involves assessing the links that the page incorporates (some links may be more important than others), determining if there are external links that use HTTP but which support HTTPS, and, assuming that such links are present, editing at least one of the HTTPs-supporting HTTP links so that the link uses HTTPS instead. The idea is that a visitor to the page will get the benefits of HTTPS even if they do not have HTTPS Everywhere installed.

The two browsers that I usually use when editing pages in this manner both have HTTPS Everywhere installed. To determine if an HTTP link supports HTTPS, one technique is to temporarily disable HTTPS Everywhere and then see if the link works when accessed via HTTPS instead of HTTP. Another technique is to use [Rex Swain's HTTP Viewer](http://www.rexswain.com/httpview.html) to see if the HTTP link URL generates a redirect (such as 301 Moved Permanently) to an HTTPS URL.

The question is, is it permissible to repeatedly manually change links on pages to use HTTPS instead of HTTP when the HTTPS Everywhere add-on has rulesets that cover the links? Or would this violate the license terms, if any, for HTTPS Everywhere and its ruleset database? (For example, HTTPS Everywhere has rulesets for Ars Technica, FanFiction.net, Goodreads.com, and Home Depot, among many other sites. If a Web page includes an HTTP link to an Ars Technica article and an HTTP link to a FanFiction.net story, and another Web page includes an HTTP link to a Goodreads.com book page and an HTTP link to a Home Depot page, would it be permissible to manually change the four affected links on those two pages to use HTTPS instead?)

**Comment by jeremyn** [comment]
@galenide I guess you are asking whether it's a copyright violation to use the HTTPS Everywhere data to change links on e.g. Wikipedia to be HTTPS? I'm not as a lawyer but AFAIK it's not clear whether something like a list of the HTTPS status of URLs is copyrightable in the U.S.: see https://en.wikipedia.org/wiki/Feist_Publications,_Inc.,_v._Rural_Telephone_Service_Co. and https://gigaom.com/2011/02/02/419-if-bing-is-copying-googles-search-results-is-it-breaking-the-law/ .


I created issue https://github.com/EFForg/https-everywhere/issues/7978 a while ago to get clarification on what HTTPS Everywhere is licensed under. I'm still not totally sure what the exact license but it seems like it must be some kind of FLOSS license, if that makes you feel better.

**Comment by jeremyn** [comment]
@galenide I guess you are asking whether it's a copyright violation to use the HTTPS Everywhere data to change links on e.g. Wikipedia to be HTTPS? I'm not as a lawyer but AFAIK it's not clear whether something like a list of the HTTPS status of URLs is copyrightable in the U.S.: see https://en.wikipedia.org/wiki/Feist_Publications,_Inc.,_v._Rural_Telephone_Service_Co. and https://gigaom.com/2011/02/02/419-if-bing-is-copying-googles-search-results-is-it-breaking-the-law/ .

Also if the EFF tried to enforce some kind of copyright claim against you for improving HTTPS coverage in the wild using their data, that would be pretty disgusting and really bad public relations.

I created issue https://github.com/EFForg/https-everywhere/issues/7978 a while ago to get clarification on what HTTPS Everywhere is licensed under. I'm still not totally sure what the exact license but it seems like it must be some kind of FLOSS license, if that makes you feel better. It would be nice to get that issue addressed ( @Hainish ) if it could also help answer questions like this.

**Comment by galenide** [comment]
Two clarifications:

* As far as seeking out HTTP links that support HTTPS and then manually editing the links to use HTTPS, the pages where I have made such changes include pages on Wikipedia, on other Wikimedia projects, and also some pages that I personally created.

* When I am assessing an HTTP link to see whether it supports HTTPS, my intent is to do the assessment in a manner that does not involve using or copying the HTTPS Everywhere ruleset database. As such, the techniques that I have used include such things as completely disabling HTTPS Everywhere in the browser and then manually accessing the link after substituting https:// for http://, or using the Web Developer functionality in Firefox to view the requests that are made when the browser accesses the link with HTTPS Everywhere disabled, or using the previously-mentioned "HTTP Viewer" from Rex Swain. In multiple cases, I have found HTTP links that support HTTPS and for which HTTPS Everywhere does not have any rulesets that cover the links. Even so, there have been times where I have manually assessed an HTTP URL as described and have found that the URL supports HTTPS, and then subsequently I find out that HTTPS Everywhere does in fact have a ruleset that covers the link in question. Or, I find an HTTP URL and I happen to know that the URL is covered or is likely covered by an HTTPS Everywhere ruleset.

As far as the question of specifically using the HTTPS Everywhere database to rewrite URLs outside of HTTPS Everywhere, the issue seems to be worth considering. In 2010, the EFF mentioned that a search engine, Duck Duck Go, was [applying HTTPS to links](https://www.eff.org/deeplinks/2010/10/search-engines-protect-privacy-outbound-https) in their search results using the HTTPS Everywhere rules. Another relevant development was that Cloudflare, a content delivery network, worked with the EFF to use the HTTPS Everywhere database to do [automatic rewriting of URLs](https://blog.cloudflare.com/fixing-the-mixed-content-problem-with-automatic-https-rewrites/) on certain pages delivered via Cloudflare.

**Comment by Bisaloo** [comment]
Related to your last comment: https://github.com/EFForg/https-everywhere/issues/6937

**Comment by jeremyn** [comment]
(Again, I can't speak for the EFF here.)

@Bisaloo 's link is very relevant I think. I would consider your manual updates as a sort of "downstream dependency".

If I were you I would have no qualms about referring to the HTTPS Everywhere database when making URL updates around the internet. I would be very irritated as a contributor if the EFF pressed some sort of claim against you for doing so.

**Comment by Hainish** [comment]
As far as I understand it, the HTTPS Everywhere rulesets themselves are not copyrightable, as they comprise a raw data set of 'facts' - the fact for instance, that one insecure URL maps to another secure URL.  The selection and arrangement of facts can be copyrightable - if it is creative enough.  The white pages of telephone books were not, but yellow pages were.

The core codebase is copyrighted under the GPLv2 license, and there are various external dependencies listed in https://github.com/EFForg/https-everywhere/blob/master/LICENSE.txt which are copyrighted other open source licenses.

At EFF we've worked very explicitly to expand the public domain and the ability for users and developers to tinker, create, and derive without fear of litigation.  For instance, all of the content on our website is under the Creative Commons Attribution license: https://www.eff.org/copyright

So there is no issue from our side using our rulesets to determine if a URL has an HTTPS endpoint.

On the remote site, it depends on their own license, I suspect.  For wikipedia, the entire purpose of the site is for it to be a collaborative, mutable information source, and it appears that it's under the Creative Commons Attribution Share-Alike license: https://en.wikipedia.org/wiki/Wikipedia:Reusing_Wikipedia_content

So there is no issue on the Wikipedia side with modifying their content to replace the insecure links with secure ones.

I can't speak to the question of whether there is ever an instance that simply changing a URL from HTTP to HTTPS will trigger a protective clause in copyright.  Even if someone wished to litigate on this edge case, it seems to me a hard sell from the litigators side.

**Comment by jeremyn** [comment]
@Hainish Thinking about HTTPS Everywhere licensing again, it occurs to me there is no such thing as a "net license" as described in [`LICENSE.txt`](https://github.com/EFForg/https-everywhere/blob/ef3d94c0b2dfdb51710d01d4dd7de2e532e305dd/LICENSE.txt). Instead the owner of the combined work chooses to release under a single license, and what that single license may be is constrained by the component licenses. Furthermore the owner may need to include the text of the component licenses depending on the terms of those licenses.

In other words it's really just waiting on the EFF to affirmatively decide to use e.g. GPLv3, or some other license that is compatible with all the components. Is there something blocking the EFF from making that decision, other than the EFF lawyers being busy with other stuff (which is understandable)?

**Comment by Hainish** [comment]
@jeremyn all that https://github.com/EFForg/https-everywhere/blob/master/LICENSE.txt is stating is that the overall license for the project is GPLv3, and that we've incorporated other code from other projects within this repo.  Some of these have their own licenses, and we explicitly state what those are.  If we were to install these libraries as dependencies rather than include them as vendorized assets, we wouldn't have to explicitly state their licenses, since the actual code for the project would not be included in this repo.

**Comment by Hainish** [comment]
I've simplified LICENSE.txt: https://github.com/EFForg/https-everywhere/commit/2481151a74ae0ea590bdb4d5dfbe8bf057606ae9

**Comment by galenide** [comment]
To clarify: In addition to certain Wikimedia projects, I have done manual assessment of HTTP links and HTTP -> HTTPS changes on a number of Web pages that I have personally authored and made available publicly at a domain name that I registered for my own usage. Most of these pages are made available under "all rights reserved" (i.e. not permissively licensed) copyright terms. For these pages that I authored, it is sometimes useful to do updates, including changing and improving the content, and fixing external links that become broken. It would seem that manually assessing and changing HTTP links to use HTTPS in such circumstances would be different from making such changes to third-party works that are copyrighted as "all rights reserved" and not permissively licensed.

If it is legally permissible to use the actual HTTPS Everywhere ruleset collection to determine, for example, whether any of the external HTTP links on a page support HTTPS, then that is useful to know, particularly with regard to other parties who may be interested in reusing the HTTPS Everywhere ruleset database to improve HTTPS coverage in other situations.
